Android TV with Zero setup

Armaxis opened the pull request:
Alternative to #1603 

Differences:
- No extra dependencies
- No initialization needed
- No `-tv` module
- Added `InternalLeakCanary.formFactor` to check what kind of setup is currently used
- Prints intent into logcat to launch `LeakActivity` for devices with launch icon bug
- Added support for new UI

Pending:
- Make banner vector drawable instead of PNG (if possible)

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/DefaultOnHeapAnalyzedListener.kt:
```diff
@@ -71,6 +89,44 @@ class DefaultOnHeapAnalyzedListener(private val application: Application) : OnHe
+      try {
+        Toast.makeText(
+            application,
+            "Analysis complete, please check Logcat",
```
- R.string?
- People using a dev app might not be developers, e.g. sales or QA. So generally it's worth providing a bit more context, at least I've tried going in that direction. You probably want something that says "Heap Analysis performed, 3 new leaks, 0 library leaks (ie copy paste the language used in the notification), see details in Logcat"

In other words tell them it's useful to see logcat.

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/DefaultOnHeapAnalyzedListener.kt:
```diff
@@ -71,6 +89,44 @@ class DefaultOnHeapAnalyzedListener(private val application: Application) : OnHe
+        )
+            .show()
+      } catch (exception: Exception) {
+        // Toasts are prone to crashing, ignore
```
What's this about? Toasts aren't prone to crashing, and if they were we should have specific code to handle those edge cases. The one crash that can happen is related to background / foreground, there are ways around it but catching will definitely do nothing as it's always crashing from a delayed message

Armaxis replied:
This comes from my old implementation of heap dump listener - I believe it was occasionally crashing after app was backgrounded - but that could've been because I was using Activity instead of Application for context. Will remove

pyricau replied:
If you do repro the crash, open up an issue that way we'll track it

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/DefaultOnHeapAnalyzedListener.kt:
```diff
@@ -71,6 +89,44 @@ class DefaultOnHeapAnalyzedListener(private val application: Application) : OnHe
+====================================
+Use following adb command to open Leak Activity with results:
+
+adb shell am start -n "${application.packageName}/${leakClass.name.replace(leakClass.simpleName, "LeakLauncherActivity")}"
```
Why not use LeakActivity directly? The two activities can be launched, no?

Armaxis replied:
It can't launch `LeakActivity` - complains about missing permission to do it. It might be because of the activity-alias?

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/DefaultOnHeapAnalyzedListener.kt:
```diff
@@ -71,6 +89,44 @@ class DefaultOnHeapAnalyzedListener(private val application: Application) : OnHe
+    val message = """====================================
+ANDROID TV LAUNCH INTENT
+====================================
+Use following adb command to open Leak Activity with results:
```
> Run the following adb command to display the list of leaks

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/LeakCanary.kt:
```diff
@@ -95,6 +95,13 @@ object LeakCanary {
+     *
+     * Defaults to [DefaultOnRetainInstanceListener]
+     */
+    val onRetainInstanceListener: OnRetainInstanceListener = DefaultOnRetainInstanceListener.create(),
```
I'd rather have us build a sealed class based API for all leak canary notifications. Until we have that, I would avoid adding more API surface that we'll have to deprecate. Instead, remove this and maybe make `DefaultOnRetainInstanceListener` and `OnRetainInstanceListener` private (or inline its behavior)

Armaxis replied:
Good point, this API is not ready to be public yet.

Armaxis replied:
Upon a closer look - `Config` is a `data class`, so we can't hide the `OnRetainInstanceListener` with `internal` modifier and keep it in the constructor. I can move it outside the constructor for now
```kotlin
data class Config(
    val dumpHeap
    ...
) {
  val onRetainInstanceListener = ..
}
```
Later, when it is ready to be public we can move it back to constructor. Does it sound good?

Side note: this would break the bytecode compatibility, but, well...

pyricau replied:
Sorry, nope: if we're not adding something as public API, then we should not be adding it to Config. Sources of API classes are read by users of the library so we should aim to limit private code in it.

Armaxis replied:
Gotcha, will work around and keep it away from public apis

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/OnRetainInstanceListener.kt:
```diff
@@ -0,0 +1,20 @@
+
+sealed class RetainInstanceChange {
+  data class CountChanged(val retainedCount: Int) : RetainInstanceChange()
+  object Reset : RetainInstanceChange()
```
Oh I see this is starting to look like the API I want, but I think it should be for all types of notifications not just "retain instance change"

Armaxis replied:
Yup, will keep it private while it's baking :)

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/DefaultOnRetainInstanceListener.kt:
```diff
@@ -0,0 +1,22 @@
+ */
+class DefaultOnRetainInstanceListener : OnRetainInstanceListener {
+
+  override fun onChange(change: RetainInstanceChange) {}
```
The default shouldn't be "empty", it should be showing notifications.

Armaxis replied:
This is  a `TODO` - move the notifications into default implementation and have TV implementations do only toasts. But it might take a while to do

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/InternalLeakCanary.kt:
```diff
@@ -247,5 +254,23 @@ internal object InternalLeakCanary : (Application) -> Unit, OnObjectRetainedList
+    MOBILE,
+    TV,
+    WATCH,
+    INSTANT_APP,
```
instant app is not a form factor. What will we do when instant apps run on watch or TV?

Armaxis replied:
I also questioned putting `INSTANT_APP` here. Even though it's mutually exclusive today doesn't guarantee it in the future.
Will separate it out

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/tv/TvOnRetainInstanceListener.kt:
```diff
@@ -0,0 +1,68 @@
+ * It will notify user when heap dump is going to happen, so that users who set the
+ * [LeakCanary.Config.retainedVisibleThreshold] to any value other than 1 will be aware of retain
+ * instances changes. If user set threshold for heap dump to 1 then this listener becomes no-op,
+ * as any retain instance change will trigger the heap dump with its own toast.
```
I think we'll want to say "hey I'm dumping the heap" (and have a default impl that's showing the toast)

Armaxis replied:
There will be a standard toast telling user that the heap is being dumped. That's why I made this one no-op for that case - so it doesn't show two toasts

pyricau commented on leakcanary-android-sample-tv/build.gradle:
```diff
@@ -0,0 +1,60 @@
+apply plugin: 'com.android.application'
+apply plugin: 'kotlin-android'
```
Could this be a flavor of the example app instead?

Armaxis replied:
I don't see why not. I can modify the regular sample app to work both on tv and mobile. Not a flavor, it will be same APK, just with extra code to run for TV.

pyricau replied:
Works for me!

Armaxis replied:
Added a button that leaks an activity (by recreating it) - it's very easy to click it on TV; works for mobile as well

pyricau replied:
>  I can modify the regular sample app

why is there still a leakcanary-android-sample-tv ?

Armaxis replied:
Oops, I need to do some clean up

Armaxis commented:
Addressed all the PR comments, please give it a look again 🙏

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/DefaultOnHeapAnalyzedListener.kt:
```diff
@@ -29,7 +35,7 @@ class DefaultOnHeapAnalyzedListener(private val application: Application) : OnHe
     }
 
-    val (contentTitle, screenToShow) = when (heapAnalysis) {
+     val (contentTitle, screenToShow) = when (heapAnalysis) {
```
?

Armaxis replied:
Lol

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/DefaultOnHeapAnalyzedListener.kt:
```diff
@@ -55,6 +73,50 @@ class DefaultOnHeapAnalyzedListener(private val application: Application) : OnHe
+   * to direct them to Logcat where a much more detailed report will be printed.
+   */
+  private fun showToast(heapAnalysis: HeapAnalysis) {
+    // Post the Toast into main thread and wrap it with try-catch in case Toast crashes (it happens)
```
This comment is out of date

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/DefaultOnHeapAnalyzedListener.kt:
```diff
@@ -55,6 +73,50 @@ class DefaultOnHeapAnalyzedListener(private val application: Application) : OnHe
+        is HeapAnalysisFailure -> application.getString(R.string.leak_canary_tv_analysis_failure)
+      }
+      Toast.makeText(
+          application,
```
See leakcanary.internal.AndroidHeapDumper#showToast

Toasting when no activity is resumed may lead to a crash.

Armaxis replied:
I gave it a look, and in the `AndroidHeapDumper` toast is being created with Activity's context. In my case I use the `Application` as a context. As a result, I am able to see Toast appear even when the app is backgrounded, which is ok for TV devices since we want to let users know that memory dump is completed when app is backgrounded.

pyricau replied:
yeaaaah but.. In recent Android versions you can no longer toast when the app is in background, hence the use of activity context and check for resume.

Armaxis replied:
Whoa, didn't know about that. It might be a problem because I don't have the reference to the activity here, need to pass it over from somewhere..
Do you have more context on that requirement about backgrounded activities? Looking at `Toast`'s source code in API 29 and it still mentions that passing Application context is ok 😟

pyricau replied:
Passing the app context might be ok but I think you must be visible? And it crashes on some version of Android (that famous toast crash) but not anymore.

Anyway `leakcanary.internal.AndroidHeapDumper#showToast` shows you how to do all that.

Armaxis replied:
Oh, so that famous crash might be the reason I had the `try-catch` block in first place. 
Cool, will try to fix it via activity context

pyricau replied:
> even when backgrounded

welp the idea is to only keep the activity reference while the activity is resumed (and therefore won't crash when toasting) so that likely won't work

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/DefaultOnHeapAnalyzedListener.kt:
```diff
@@ -55,6 +73,50 @@ class DefaultOnHeapAnalyzedListener(private val application: Application) : OnHe
+   */
+  private fun printIntentInfo() {
+    val leakClass = LeakActivity::class.java
+    val message = """====================================
```
Log related code should execute within the log lambda

Armaxis replied:
Oooh, good catch!

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/DefaultOnHeapAnalyzedListener.kt:
```diff
@@ -55,6 +73,50 @@ class DefaultOnHeapAnalyzedListener(private val application: Application) : OnHe
+====================================
+Run the following adb command to display the list of leaks:
+
+adb shell am start -n "${application.packageName}/${leakClass.name.replace(leakClass.simpleName, "LeakLauncherActivity")}"
```
`leakClass.name.replace(leakClass.simpleName, "LeakLauncherActivity")` 

Isn't that `leakClass.package.name + "LeakLauncherActivity"` ?

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/HeapDumpTrigger.kt:
```diff
@@ -102,6 +105,8 @@ internal class HeapDumpTrigger(
       retainedReferenceCount = objectWatcher.retainedObjectCount
     }
 
+    onRetainInstanceListener.onChange(CountChanged(retainedReferenceCount))
```
did the count actually change?

Armaxis replied:
Not necessarily, it might be the same count and it is fine. It's used to let listener know that we just performed a retained instances check and the parameter is the current `retainedReferenceCount`.
The `CountChanged` is incorrect name, I can rename it to `MaybeCountChanged` or `CountUpdated`. Word `updated` doesn't suggest it actually changed, right?

pyricau replied:
How about only sending CountChanged when the count actually changed?

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/InternalLeakCanary.kt:
```diff
@@ -76,6 +83,33 @@ internal object InternalLeakCanary : (Application) -> Unit, OnObjectRetainedList
+
+  val formFactor by lazy {
+    val currentModeType =
+      (application.getSystemService(UI_MODE_SERVICE) as UiModeManager).currentModeType
```
wow, UiModeManager has been here since API lvl 8, first time I see it.

Armaxis replied:
I felt the same way :)

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/InternalLeakCanary.kt:
```diff
@@ -76,6 +83,33 @@ internal object InternalLeakCanary : (Application) -> Unit, OnObjectRetainedList
     )
 
+  internal enum class FormFactor {
+    MOBILE,
```
So tablets are MOBILE :) ?

Armaxis replied:
They are, but I share your confusion here! I think we need a better name that will cover both mobile and tablet. Is there a commonly accepted term to describe these two?

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/OnRetainInstanceListener.kt:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Listener set in [LeakCanary.Config] and called by LeakCanary on a background thread when
+ * the number of retained instances changes.
+ *
```
javadoc not up to date. Also, empty line.

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/tv/TvOnRetainInstanceListener.kt:
```diff
@@ -0,0 +1,69 @@
+        Toast.makeText(application, text, Toast.LENGTH_LONG).show()
+      } catch (exception: Exception) {
+        // Toasts are prone to crashing, ignore
+      }
```
ah, the try catch is back

Armaxis replied:
Nah, I forgot that I had this here as well, shame on me.

pyricau commented on leakcanary-android-sample/src/main/AndroidManifest.xml:
```diff
@@ -19,6 +19,13 @@
+      android:required="false" />
+  <uses-feature
+      android:name="android.hardware.touchscreen"
+      android:required="false" />
```
Why is this here?

Armaxis replied:
These two are required for the app to be runnable on TV. 
I never tried running without these, though. Wonder if 'required' is not that strict of a requirement. Will try removing!

Armaxis replied:
If these are removed - the manifest starts to complain about missing features; although the app would still run on TV. They are needed, however, to make the app appear in the Play Store as available for TV:
https://developer.android.com/training/tv/start/start.html#leanback-req
So, while not necessary for our sample app, it's safe to keep it to make Lint happy.

What's more interesting is that I was wrong by adding these features to the `leakcanary-android-core`'s `AndroidManifest.xml`. Library doesn't have to define it, it's up to the users of the library to add TV support in their apps. I removed the features from that manifest file

pyricau replied:
Nice! Ok I was confused as to why it was added in both places. That way is even better :)

pyricau commented on leakcanary-android-sample/src/main/res/layout/main_activity.xml:
```diff
@@ -29,5 +29,12 @@
       android:text="@string/helper_text"
       />
+  <Button
+      android:id="@+id/button"
```
better id name?

pyricau commented on leakcanary-android-sample/src/main/res/values/strings.xml:
```diff
@@ -18,7 +18,8 @@
   <string name="app_name">Example App</string>
-  <string name="helper_text">Rotate the screen 5 times to create 5 leaks and trigger a foreground
-    heap dump. Or press back to cause a leak and trigger a background heap dump.
+  <string name="helper_text">Rotate the screen or press the button 5 times to create 5 leaks and
```
Should be clear that pressing the button will trigger an activity recreation

Armaxis commented:
Addressed PR comments, added vector icon.
You can review per-commit:
c881cf685a1a72a54c139a33488888817b86efae
216d41766e16f02945da04ad2327c1c29cafd1f8
beb9bdd4efd22e58a1264d08b970ff1b71be9df7

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/DefaultOnHeapAnalyzedListener.kt:
```diff
@@ -21,6 +26,7 @@ import shark.SharkLog
  * show a notification summarizing the result.
  */
 class DefaultOnHeapAnalyzedListener(private val application: Application) : OnHeapAnalyzedListener {
+  private val handler = Handler(Looper.getMainLooper())
```
nit: "mainHandler" or "uiHandler"

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/DefaultOnHeapAnalyzedListener.kt:
```diff
@@ -55,6 +73,55 @@ class DefaultOnHeapAnalyzedListener(private val application: Application) : OnHe
+    handler.post {
+      val resumedActivity = InternalLeakCanary.resumedActivity
+      if (resumedActivity == null) {
+        SharkLog.d { "Can't display Toast, activity is backgrounded" }
```
Devs have no idea what toast we're trying to display. Not sure it's worth logging, but if we do we should be more specific.

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/DefaultOnHeapAnalyzedListener.kt:
```diff
@@ -55,6 +73,55 @@ class DefaultOnHeapAnalyzedListener(private val application: Application) : OnHe
+  }
+
+  /**
+   * Android TV with API 26+ has a bug where launcher icon doesn't appear, so users won't know how
```
"where launcher icon" => "where the launcher icon"

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/DefaultOnHeapAnalyzedListener.kt:
```diff
@@ -55,6 +73,55 @@ class DefaultOnHeapAnalyzedListener(private val application: Application) : OnHe
+  /**
+   * Android TV with API 26+ has a bug where launcher icon doesn't appear, so users won't know how
+   * to launch LeakCanary Activity.
+   * This method would print an adb command that launched LeakCanary into the logcat
```
`This method prints`

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/tv/TvOnRetainInstanceListener.kt:
```diff
@@ -0,0 +1,64 @@
+        application.getString(R.string.leak_canary_notification_retained_dump_wait)
+      }
+    }
+    displayToast("LeakCanary: $message")
```
bird icon!

Armaxis replied:
Oooh, didn't think of that!

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/tv/TvOnRetainInstanceListener.kt:
```diff
@@ -0,0 +1,64 @@
+    handler.post {
+      val resumedActivity = InternalLeakCanary.resumedActivity
+      if (resumedActivity == null) {
+        SharkLog.d { "Can't display Toast, activity is backgrounded" }
```
This log isn't super clear without context.

pyricau commented on leakcanary-android-sample/src/main/res/values/strings.xml:
```diff
@@ -18,7 +18,9 @@
-    heap dump. Or press back to cause a leak and trigger a background heap dump.
+  <string name="helper_text"> - Rotate the screen 5 times to create 5 leaks and trigger a foreground heap dump\n
+    - Press the button to recreate activity 5 times to create 5 leaks and trigger a foreground heap dump\n
+    - Press back to cause a leak and trigger a background heap dump
```
Pressing back does not cause a leak, it changes the threshold and triggers the heap dump.

We need to explain that either rotating or pressing the button will create 1 leak. That's it. Then we also need to explain that after 5 leaks the analysis will trigger.

...

Or maybe we don't, this is a sample app that generally most people won't even see.

Armaxis replied:
I spent good 10 minutes thinking about properly wording it and still can't figure it out. I'll file an issue to work on this sample app text.

Armaxis replied:
On the second thought, take a look at new version..

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/DefaultOnHeapAnalyzedListener.kt:
```diff
@@ -55,6 +73,55 @@ class DefaultOnHeapAnalyzedListener(private val application: Application) : OnHe
+          val leakTypeCount = heapAnalysis.applicationLeaks.size + heapAnalysis.libraryLeaks.size
+          application.getString(
+              R.string.leak_canary_tv_analysis_success,
+              leakTypeCount,
```
this should be `heapAnalysis.applicationLeaks.size`

Shilpa2212 commented:
@pyricau Seems this has been approved, when will it get merged?
Was looking since we need to check TV application for memory leaks

Armaxis commented:
@Shilpa2212 everything except changes in last commit was approved. Once that one is approved as well we will proceed with fixing it. Current target is to shop TV support with next LeakCanary release

pyricau commented:
Damn, I did a merge instead of a squash, so all those 21 commits are now on master. Oh well. Next time, can you squash as you go ;) ?

Armaxis commented:
@pyricau I kept it unsquashed so that you can review changes on last commit only and intended to squash during merge :) 
Oh well, we have to live with this now :)
